### PR TITLE
[4.0] State to stage fixes

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/21553#issuecomment-412333785

### Summary of Changes
Some more fixes in this renaming task.
Rename the Field WorkfileState to Workfilestage

Thank you for all hints and for testing.

### CODE REVIEW

**Comment by HLeithner** [comment]
There is a typo at
https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_workflow/Model/TransitionsModel.php#L143-L144

The semicolon show be in the second line

**Comment by HLeithner** [comment]
Also there seams to be missing string translations/renamings in transition
image

**Comment by HLeithner** [comment]
https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_workflow/Table/StageTable.php#L69
should be $stage

File: modules/mod_articles_archive/mod_articles_archive.xml [hunk 1]
```diff
@@ -32,7 +32,7 @@
 
 				<field
 					name="state"
-					type="workflowstate"
+					type="workflowstage"
 					multiple="true"
 					size="5"
 					label="MOD_ARTICLES_ARCHIVE_CHOOSE_STATE"
```
Comment (jeckodevelopment) [inline modules/mod_articles_archive/mod_articles_archive.xml:38]: `MOD_ARTICLES_ARCHIVE_CHOOSE_STATE` should be `MOD_ARTICLES_ARCHIVE_CHOOSE_STAGE`

**Review by jeckodevelopment — CHANGES_REQUESTED** [review]

File: modules/mod_articles_archive/mod_articles_archive.xml (hunk 1 above)
Comment (chmst) [reply modules/mod_articles_archive/mod_articles_archive.xml:38]: thx

**Comment by HLeithner** [comment]
In the SQL update script you changed the last line

```sql
SELECT `id`, CASE WHEN `stage` = -2 THEN 3 WHEN `stage` = 0 THEN 1 WHEN `stage` = 2 THEN 4 ELSE 2 END, 'com_content' FROM `#__content`;
```

The stage column here references the state column in #__content so it have to stay `state`

**Comment by HLeithner** [comment]
JHtmlWorkflowStage is missing, file and class is in lib/cms/html/workflowstate.php

**Comment by HLeithner** [comment]
oh lib/cms/html/workflowstate.php only got converted partly.

In line 62 wrong name
```php
		$workflowStates = array();
```
in line 70 the first variable is wrong
```php
			if (!array_key_exists($workflowStateKey, $workflowStages))
```

**Comment by chmst** [comment]
This file should have been replaced already by workflowstage. I'll Repeating the changes. Thank you.

**Comment by HLeithner** [comment]
libraries/src/Form/Field/TransitionField.php

workflow_state is wrong, also in the query some lines later
```php
$workflowState = $this->element['workflow_state'] ? (int) $this->element['workflow_state'] : (int) $jinput->getInt('id', 0);
```

**Comment by HLeithner** [comment]
administrator/components/com_content/Model/ArticleModel.php line 545 is crying Notice: Undefined property: stdClass::$state_id 

also it sets 'workflow_state' not sure if this is correct but I don't think so.

**Comment by HLeithner** [comment]
administrator/components/com_content/Model/ArticleModel.php:827

```php
					->where($db->quoteName('wt.to_state_id') . ' = ' . $db->quoteName('ws.id'))
```
is wrong

**Comment by HLeithner** [comment]
administrator/components/com_content/Model/ArticleModel.php:913
```
					$stateId = (int) $workflow->state_id;
```
some lines later:
```
			// If we have a new state, create the workflow association
			if (!empty($stateId))
			{
				$workflow->createAssociation($this->getState($this->getName() . '.id'), (int) $stateId);
			}
```

administrator/components/com_content/Model/ArticleModel.php:1222

```
					->select($db->quoteName('ws.id', 'state_id'))
```

**Comment by wilsonge** [comment]
You still can't save an article with this PR on :) - that's due to this one https://github.com/joomla/joomla-cms/pull/21566#issuecomment-412369405 i think

**Comment by infograf768** [comment]
Indeed
`Call to undefined method stdClass::createAssociation()`

**Comment by infograf768** [comment]
Checked: remain many occurences of `$workflowState` instead of `$workflowStage`, including the name of the file
`/libraries/cms/html/workflowstate.php` instead of `workflowstage.php`

EDIT: for the name of the file, it looks it is done in the PR. (Eclipse can't move files when applying patches.)

Also `$stateId` instead of `$stageId`

After correcting these I can save articles here.

**Comment by infograf768** [comment]
This may help correcting $workflowState
```
/administrator/components/com_content/Model/ArticlesModel.php:322: 		$workflowState = (string) $this->getState('filter.state');
/administrator/components/com_content/Model/ArticlesModel.php:324: 		if (is_numeric($workflowState))
/administrator/components/com_content/Model/ArticlesModel.php:326: 			$query->where('wa.stage_id = ' . (int) $workflowState);
/administrator/components/com_content/Model/ArticlesModel.php:341: 		elseif (!is_numeric($workflowState))
/administrator/components/com_content/Model/FeaturedModel.php:160: 		$workflowState = (string) $this->getState('filter.state');
/administrator/components/com_content/Model/FeaturedModel.php:162: 		if (is_numeric($workflowState))
/administrator/components/com_content/Model/FeaturedModel.php:164: 			$query->where('wa.stage_id = ' . (int) $workflowState);
/administrator/components/com_content/Model/FeaturedModel.php:173: 		elseif (!is_numeric($workflowState))
```
and for `$stateId`
```
/administrator/components/com_content/Model/ArticleModel.php:891: 			if (empty($stateId))
/administrator/components/com_content/Model/ArticleModel.php:924: 				$workflow->createAssociation($this->getState($this->getName() . '.id'), (int) $stateId);
```

File: administrator/components/com_content/Model/FeaturedModel.php [hunk 2]
```diff
@@ -156,12 +156,12 @@ protected function getListQuery()
 			$query->where('c.access IN (' . $groups . ')');
 		}
 
-		// Filter by published state
-		$workflowState = (string) $this->getState('filter.state');
+		// Filter by workflows stages
+		$workflowStage = (string) $this->getStage('filter.state');
```
Comment (infograf768) [inline administrator/components/com_content/Model/FeaturedModel.php]: getStage ?

Comment (chmst) [reply administrator/components/com_content/Model/FeaturedModel.php]: oops

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql [hunk 3]
```diff
@@ -120,5 +120,5 @@ INSERT INTO `#__extensions` (`extension_id`, `package_id`, `name`, `type`, `elem
 --
 -- Creating Associations for existing content
 --
-INSERT INTO `#__workflow_associations` (`item_id`, `stage_id`, `extension`)
-SELECT `id`, CASE WHEN `stage` = -2 THEN 3 WHEN `stage` = 0 THEN 1 WHEN `stage` = 2 THEN 4 ELSE 2 END, 'com_content' FROM `#__content`;
+INSERT INTO `#__workflow_associations` (`item_id`, `state_id`, `extension`)
```
Comment (infograf768) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql]: I guess it is `stage_id`
in `INSERT INTO `#__workflow_associations` (`item_id`, `state_id`, `extension`)`

Comment (chmst) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql]: thx

**Comment by infograf768** [comment]
Are Default transitions supposed to be this?
screen shot 2018-08-13 at 17 16 47

If yes, this PR is fine for me (have not tested all possibilities, but no major error.)

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 9d4458c9100afa7a785189c61415eeb60898ea3d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21566">issues.joomla.org/tracker/joomla-cms/21566</a>.</sub>

**Comment by wilsonge** [comment]
Thanks for all the great collaboration here guys :) much appreciated!

**Comment by HLeithner** [comment]
I have tested this item :white_check_mark: successfully on 9d4458c9100afa7a785189c61415eeb60898ea3d<br /><br />No errors found so far.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21566">issues.joomla.org/tracker/joomla-cms/21566</a>.</sub>

**Comment by brianteeman** [comment]
complete failure for me with 500 error as soon as you save an article

**Comment by HLeithner** [comment]
do you have more debug information? for me this works perfectly :-(

**Comment by brianteeman** [comment]
As soon as you create an article you get this everytime you try to go to 
administrator/index.php?option=com_content

Error 500:  JHtml workflowstage not found.
```

# | Function | Location
-- | -- | --
1 | () | JROOT\libraries\src\HTML\HTMLHelper.php:164
2 | Joomla\CMS\HTML\HTMLHelper::_() | JROOT\layouts\joomla\html\batch\workflowstage.php:31
3 | include() | JROOT\libraries\src\Layout\FileLayout.php:127
4 | Joomla\CMS\Layout\FileLayout->render() | JROOT\libraries\src\Layout\LayoutHelper.php:73
5 | Joomla\CMS\Layout\LayoutHelper::render() | JROOT\administrator\components\com_content\tmpl\articles\default_batch_body.php:49
6 | include() | JROOT\libraries\src\MVC\View\HtmlView.php:386
7 | Joomla\CMS\MVC\View\HtmlView->loadTemplate() | JROOT\administrator\components\com_content\tmpl\articles\default.php:338
8 | include() | JROOT\libraries\src\MVC\View\HtmlView.php:386
9 | Joomla\CMS\MVC\View\HtmlView->loadTemplate() | JROOT\libraries\src\MVC\View\HtmlView.php:191
10 | Joomla\CMS\MVC\View\HtmlView->display() | JROOT\administrator\components\com_content\View\Articles\HtmlView.php:149
11 | Joomla\Component\Content\Administrator\View\Articles\HtmlView->display() | JROOT\libraries\src\MVC\Controller\BaseController.php:679
12 | Joomla\CMS\MVC\Controller\BaseController->display() | JROOT\administrator\components\com_content\Controller\DisplayController.php:59
13 | Joomla\Component\Content\Administrator\Controller\DisplayController->display() | JROOT\libraries\src\MVC\Controller\BaseController.php:717
14 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT\libraries\src\Dispatcher\Dispatcher.php:162
15 | Joomla\CMS\Dispatcher\Dispatcher->dispatch() | JROOT\libraries\src\Component\ComponentHelper.php:354
16 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT\libraries\src\Application\AdministratorApplication.php:111
17 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT\libraries\src\Application\AdministratorApplication.php:154
18 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT\libraries\src\Application\CMSApplication.php:242
19 | Joomla\CMS\Application\CMSApplication->execute() | JROOT\administrator\includes\app.php:63
20 | require_once() | JROOT\administrator\index.php:36


```

**Comment by mbabker** [comment]
https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/cms/html/workflowstage.php

This is environmentally broken.

Because this uses the Joomla autoloader, which is based on camel casing of class names, the class name MUST be `JHtmlWorkflowstage` otherwise it will NOT work consistently across environments.

OR, to keep the class casing the way it is, you must place the file at `libraries/cms/html/workflow/stage.php`.

**Comment by HLeithner** [comment]
@mbabker but this doesn't seams to be the problem because the file called JHtml key is 'workflowstage.existing' and the file is also lowercase only the class is SnakeCase so whats the reason it works for me and not for brian?

**Comment by mbabker** [comment]
And when (both of) you step through all the autoloaders and whatnot what path and class name are trying to be resolved?  This is PHP here, it is very well possible for the right file to be loaded in but the class still be considered missing (which is basically what this error is) because the JHtml library is expecting to find `JHtmlWorkflowstage` and the class is declared `JHtmlWorkflowStage`.

**Comment by HLeithner** [comment]
So this should be fixed by this https://github.com/joomla/joomla-cms/pull/21597 PR.

**Pull request merged**